sw/device/lib/dif/templates/dif_autogen.c.tpl

```diff
@@ -0,0 +1,211 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+// This file is auto-generated.
+
+#include "sw/device/lib/dif/dif_${ip_snake}_autogen.h"
+
+#include "${ip_snake}_regs.h"  // Generated.
+
+/**
+ * Get the corresponding interrupt register bit offset. INTR_STATE,
+ * INTR_ENABLE and INTR_TEST registers have the same bit offsets, so this
+ * routine can be reused.
+ */
+static bool ${ip_snake}_get_irq_bit_index(
+  dif_${ip_snake}_irq_t irq,
+  bitfield_bit32_index_t *index_out) {
+
+  switch (irq) {
+${irq_to_bit_index_cases}
+    default:
+      return false;
+  }
+
+  return true;
+}
+
+DIF_WARN_UNUSED_RESULT
+dif_${ip_snake}_result_t dif_${ip_snake}_irq_get_state(
+  const dif_${ip_snake}_t *${ip_snake},
+  dif_${ip_snake}_irq_state_snapshot_t *snapshot) {
+
+  if (${ip_snake} == NULL || snapshot == NULL) {
+    return kDif${ip_camel}BadArg;
+  }
+
+  *snapshot = mmio_region_read32(
+        ${ip_snake}->params.base_addr, 
+        ${ip_upper}_INTR_STATE_REG_OFFSET);
```

Comment:
Optional: Maybe it's worth (for readability) using Make `<%def>`s to hoist some of these names out as variables? E.g. it might be nicer to just write `${ip_state_reg}` and whatnot. Up to you.

Reply:
Fixed.